### PR TITLE
Bounds parsing refactoring

### PR DESCRIPTION
**Pull request by matklad**
A follow up of #37278.

`parse_opt_ty_param_bounds` is introduced to make sure that we don't parse empty bonds. This fixes a couple of bugs along the way, where the check was missing. This is [breaking-change].

### CODE REVIEW

**Comment by bluss** [comment]
This kind of change would normally first have warnings, then enable the breaking change later.

**Comment by matklad** [comment]
Yep, a warning is a good idea. But perhaps we need a crater run first, to estimate the amount of breakage just in case? Or is crater capable of reporting a diff in warnings?

**Comment by bluss** [comment]
A deny by default warning works well just for crater (how it was done in rust-lang/rust/pull/37378 ). That also helps to only catch the actual crates that regress.

**Comment by matklad** [comment]
And can I emit a deny by default warning directly from the parser? I believe I can't do this in `ast_validation.rs` in a straightforward way, because it's impossible to distinguish absence of bounds from the empty bounds, both case produce AST of the same shape. I think I can look a the spans and try to determine if bounds are empty, but `:` is present, but this feels a bit too hacky. 


An alternative which I do know should work is to run crater with `span.error` and then change to `span.warn`. `warn_missing_semicolon` produces a similar warning now.

**Comment by matklad** [comment]
@pnkfelix ping :)

**Comment by alexcrichton** [comment]
@jseyfried  you may also be able to take a look and help with review

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -4186,11 +4175,25 @@ impl<'a> Parser<'a> {
...
+
+    fn parse_opt_ty_param_bounds(&mut self,
+                                 intoducing_token: &token::Token,
+                                 mode: BoundParsingMode)
+                                 -> PResult<'a, Option<TyParamBounds>> {
+        if !self.eat(intoducing_token) {
+            return Ok(None);
         }
+
+        let bounds = self.parse_ty_param_bounds(mode)?;
```
Comment (jseyfried) [inline src/libsyntax/parse/parser.rs]: @matklad `parser.eat(&token::Ident(keywords::Impl.ident()))` should do the trick.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -4478,46 +4481,40 @@ impl<'a> Parser<'a> {
...
-                        let hi = self.prev_span.hi;
-                        let span = mk_sp(lo, hi);
-
-                        if bounds.is_empty() {
-                            self.span_err(span,
-                                          "each predicate in a `where` clause must have \
-                                           at least one bound in it");
-                        }
+                    if let Some(bounds) = self.parse_opt_ty_param_bounds(&token::Colon,
+                                                                         BoundParsingMode::Bare)? {
```
Comment (jseyfried) [inline src/libsyntax/parse/parser.rs]: If you refactor this to
```rust
let opt_bounds = self.parse_opt_ty_param_bounds(&token::Colon, BoundParsingMode::Bare)?;
if let Some(bounds) = opt_bounds {
```
then you can continue using `else if` below.

**Comment by jseyfried** [comment]
If I understand correctly, there are three separate [breaking-change]s here:
 - `where 'a, ...` is no longer allowed. I believe it equivalent to `where ...` today.
 - `where 'a 'b, ...` is no longer allowed. Today, it is equivalent to `where 'a: 'b, ...`.
 - We require a type parameter bound after `:` or `+` (in the appropriate context). Today, a bound is not required in the two cases from [empty-type-parameter-bounds.rs](https://github.com/rust-lang/rust/pull/37511/files#diff-fa6e476282c88a3f65c63e32694093a1).

Could someone do a Crater run to determine which of these changes need a warning cycle?
cc @eddyb

**Comment by petrochenkov** [comment]
>We require a type parameter bound after : or +

I haven't read the patch, but always requiring a type parameter after `+` seems incorrect, it's a usual case of trailing separator, they are always allowed <sub><sup>(except for trailing `|` in `match` arms)</sup></sub>.

**Comment by jseyfried** [comment]
I think of `+` not as a separator but as a associative operator -- e.g. `let x = 1 +;` wouldn't make sense.

**Comment by eddyb** [comment]
Started crater run (sorry for the delay).

**Comment by eddyb** [comment]
[Crater report](https://gist.github.com/eddyb/8a2d26d2f42d2d991c197e44eb844d7d) shows 8 regressions (only `eetf` is a false positive).
Most of them have something like `trait Provider: {` which is just weird.
*However*, the ones which generate empty lists of bounds through macros seem legitimate.

**Comment by jseyfried** [comment]
@eddyb I think a macro generating empty lists of bounds could have a legitimate use case, but it looks like the macro-expanded bounds from the Crater run are always empty and can be easily avoided. Also, we already forbid empty lists of bounds in other contexts today.

@matklad The all the breakage appears to due to empty bounds lists after `:`. Could you make this a warning instead of a hard error?

**Comment by petrochenkov** [comment]
I've audited `ast.rs` and it turns out there are very few lists where empty list is syntactically invalid.

1. Path segments. Empty path may make sense semantically - current module if it's local or crate root if it's global, but there are obvious parsing problems.
2. Patterns in match arm. An match arm without patterns (impossible arm) doesn't make sense semantically. 
3. Type or lifetime bounds. Empty list is semantically valid (no bounds) and may be generated by macros.
4. Where predicates. Empty list is semantically valid (no predicates) and may be generated by macros.

Basically, I think this should be valid, even if it looks somewhat weird:
```
struct S<T:> where {
    field: T,
}
```

**Comment by petrochenkov** [comment]
@jseyfried 
>I think of + not as a separator but as a associative operator -- e.g. let x = 1 +; wouldn't make sense.

I admit it may be interpreted as sum in type sums `type T = Trait + Send;`, but it's certainly a list separator in bounds.
If both cases are parsed by the same routine now, then I don't mind prohibiting trailing `+`, it's not a big deal and can always be allowed later if convincing use case is found.

**Comment by matklad** [comment]
Looks like it is not clear if empty parameter bounds should be allowed. So I've resurrected #37278 which only fixes a clear bug with lifetimes in the where clause. 

So I think we need to tag this with T-lang and decide what syntax should be allowed. Here is the status quo:

```Rust
// These are accepted
trait A: {}
fn b<'a:, U:>() {}
type C = for<'a> Clone+;

// These are forbidden

// bounds on where clauses must be non empty
fn d<T>() where T: {}

// In type grammar, `+` is treated like a binary operator,
// and hence both L and R side are required.
fn e(f: &(A+)) {}
```

**Comment by jseyfried** [comment]
I'm leaning toward @petrochenkov's [proposal](https://github.com/rust-lang/rust/pull/37511#issuecomment-260127091) to allow empty bounds lists (e.g. `fn f<T:>() {}`).
I don't have a strong opinion trailing `+`s, which aren't as important since we can forbid them consistently  without without breakage in practice and decide later.

@rust-lang/lang do we want to allow empty bounds lists and/or trailing `+`s? Both are [sometimes](https://github.com/rust-lang/rust/pull/37511/files#diff-fa6e476282c88a3f65c63e32694093a1) allowed today. Forbidding empty bounds lists would cause some trivially fixable [breakage](https://gist.github.com/eddyb/8a2d26d2f42d2d991c197e44eb844d7d), so we would need a warning cycle.

**Comment by nrc** [comment]
IMO we should allow neither empty bounds lists nor trailing `+` both are weird. However, I'm somewhat sympathetic to the macro/codegen argument for the bounds list case.

**Comment by matklad** [comment]
Worth mentioning that currently `where` clauses allow empty bounds for lifetimes, but not for types. 

Valid: `where 'a:, T: Foo`

Invalid: `where 'a: 'b, T:`

**Comment by nikomatsakis** [comment]
At the @rust-lang/lang meeting recently we settled on:

- accept empty bounds list for both types/lifetime parameters
- do not accept trailing `+` for now because @nrc thinks it looks weird
    - macros can instead generate multiple where-clauses:
        - `T: Foo, T: Bar,`

Sound good?

**Comment by matklad** [comment]
> Sound good?

Yep. Stuff like `trait Provider: {` can be cleaned up by IDE / rustfmt.

**Comment by pnkfelix** [comment]
@matklad does the current PR encode that semantics yet, or does it need further revision?

**Comment by matklad** [comment]
@pnkfelix yep, this does not implement [the suggestion](https://github.com/rust-lang/rust/pull/37511#issuecomment-264312414) yet. I'll try to look into that, thanks for the reminder!

**Comment by matklad** [comment]
>I'll try to look into that, thanks for the reminder!

Looks like I don't have enough free time after all to actually make a fix :(

**Comment by petrochenkov** [comment]
@matklad 
I can rebase this and finish the work if you want (not right now though, but in a couple of weeks).

**Comment by matklad** [comment]
@petrochenkov that'd be great! Not sure that rebasing makes sense though: the current implementation is rather different from the final proposal we arrived at!

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/39158 is submitted.

**Comment by aturon** [comment]
Should we close this PR in favor of #39158?

**Comment by matklad** [comment]
Surely!
